Correction of Reversed Interzone Construction Check for Asymmetric Material Layers

RKStrand opened the pull request:
Pull request overview
---------------------
Most material layers are homogeneous and have the same properties throughout.  Constructions that consist of these layers are easy to check when they are used for interzone surfaces because the layers just have to be matched in reverse order.  For certain materials that have properties that differ on the front and back side, a new material layer has to be created that has these front and back side properties reversed to make sure that the construction is in fact a reverse of the original construction.  This enhanced check was not done previously, leading to error messages when the user did this properly and also no error messages when in fact there should have been a problem because one construction was not the mirror of its interzone pair.  The solution to this problem includes additional check for glass materials that have front and back side properties that are different.  This resolves GitHub Issue #4397.

### Work Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
   - Defect: This pull request repairs a github defect issue #4397

### Review Checklist
This will not be exhaustively relevant to every PR.
 - [x] Functional code review (it has to work!)
 - [x] If defect, results of running current develop vs this branch should exhibit the fix
 - [ ] CI status: all green or justified
 - [x] Performance: CI Linux results include performance check
 - [x] Unit Test(s)
 - C++ checks:
   - [x] Argument types
   - [ ] If any virtual classes, ensure virtual destructor included, other things
 - IDD changes: n/a
 - [x] Required documentation updates?

RKStrand committed:
Merge branch 'develop' into 4397-AsymPropsIZSurfaceCheck

RKStrand committed:
Merge branch 'develop' into 4397-AsymPropsIZSurfaceCheck

RKStrand committed:
Solution to Reversed Layers for IZ Constructions

When interzone surfaces are used, the constructions must be reversed.
For most materials, the layers simply need to be matched in reverse
order.  However, for window glass and equivalent glass layers, some of
the properties are different on the front and back side.  So, when this
check happens and the layers are different, it needs to see if the user
has correctly reversed these material types.  In V9.1, no check is made
and an error is thrown if the materials are different even if they are
correctly reversed.

RKStrand committed:
Merge branch 'develop' into 4397-AsymPropsIZSurfaceCheck

RKStrand committed:
Final fixes, unit test, and documentation addition

Completed some corrections to the issue solution as well as added a
unit test and some documentation to clarify the requirements for
interzone constructions.  This is now ready for review.

RKStrand committed:
Correction for build fail and removal of unused variable

The new subroutine/function was not logged on a whitelist and was
causing a build issue for the custom check compiler.  In addition, an
unused variable was removed from SurfaceGeometry.cc.

RKStrand committed:
Correction of a few issues that caused warnings

Misassigned a boolean variable as a real.  Corrected to eliminate build
warnings.

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into 4397-AsymPropsIZSurfaceCheck

mjwitte commented on src/EnergyPlus/SurfaceGeometry.cc:
```diff
@@ -13264,6 +13260,92 @@ namespace SurfaceGeometry {
     }
+    
+
+    void CheckForReversedLayers(bool &RevLayerDiffs,    // true when differences are discovered in interzone constructions
```
Would prefer for this to be a bool function and drop the first argument, but only if there are other changes needed.

RKStrand replied:
@mjwitte Yeah, had done that early on and ran into some sort of weird problems.  I don't remember what the issue was anymore, but doing this resolved the issue.  My preference would have been a bool function, too.  I'll see what shakes out with the other comments...

RKStrand replied:
@mjwitte Now I remember...it was some weird complaint about setting the value of a bool function and then resetting it.  I also found the logic here a bit more complex than a simple set it and return immediately.  So, I did this because of the complexity.  I'd like to leave this as is if you don't mind.  I will fix a couple of the other things though (see other comments).

mjwitte commented on src/EnergyPlus/SurfaceGeometry.cc:
```diff
@@ -13264,6 +13260,92 @@ namespace SurfaceGeometry {
+        int LayerNo;
+        RevLayerDiffs = false;
+        
+        for (LayerNo = 1; LayerNo <= TotalLayers; ++LayerNo) {
```
Again, nitpicking, prefer to declare the int here and delete it above:
for (int LayerNo = 1 . . . .

RKStrand replied:
Will fix this.

mjwitte commented on src/EnergyPlus/SurfaceGeometry.cc:
```diff
@@ -13264,6 +13260,92 @@ namespace SurfaceGeometry {
+                } else if ((thisMatLay.Group == GlassEquivalentLayer) && (revMatLay.Group == GlassEquivalentLayer)) {
+                    auto &thisMatLay(Material(thisConstLayer));
+                    auto &revMatLay(Material(revConstLayer));
+                    Real64 const SmallDiff = 0.0001;
```
all three of these appear to be duplicate declarations?

RKStrand replied:
@mjwitte Hmmm...looks like I forgot to eliminate these declarations after moving them up a level.  I will get rid of those.

mjwitte commented on src/EnergyPlus/SurfaceGeometry.cc:
```diff
@@ -13264,6 +13260,92 @@ namespace SurfaceGeometry {
+                (thisMatLay.Group == WindowGlass) ||    // Not window glass or glass equivalent layer which have
+                (revMatLay.Group == WindowGlass) ||     // to have certain properties flipped from front to back
+                (thisMatLay.Group == GlassEquivalentLayer) ||
+                (revMatLay.Group == GlassEquivalentLayer)) {
```
These long ORs were hard for me to follow. I think I understand them now. Not sure how to suggest making this more clear, however.

RKStrand replied:
The problem here is that there are a lot of input parameters that need to be checked and some have to agree with the reverse side.  I agree that these are pretty complex.  I don't know how to make them less complex since all of those checks have to be made.

mjwitte commented on tst/EnergyPlus/unit/SurfaceGeometry.unit.cc:
```diff
@@ -4432,3 +4432,158 @@ TEST_F(EnergyPlusFixture, SurfaceGeometry_CreateInternalMassSurfaces)
+    RevLayerDiffs = true;
+    ExpectResult = false;
+    CheckForReversedLayers(RevLayerDiffs, 1, 2, 3);
+    EXPECT_EQ(RevLayerDiffs, ExpectResult);
```
Simpler to use EXPECT_FALSE(RevLayerDiffs);

RKStrand replied:
My bad...I should have used the EXPECT_TRUE and EXPECT_FALSE functions rather than these.

mjwitte reviewed (commented):
Tested this with the user defect file, and with some variations adding more layers. The checks are working as expected.
I've noted a few nit-picky code things, but the duplicate declarations should be fixed for sure.
I've merged in develop and resolved the conflict and pushed that up to github.
And the unit tests all run ok serially.

RKStrand committed:
Changes based on code review

A few minor changes were made in response to the code review.
Hopefully this addresses all of the issues and this is ready for
merging.

RKStrand commented:
Thanks, Mike!

On Sep 18, 2019, at 8:50 PM, Michael J. Witte <notifications@github.com<mailto:notifications@github.com>> wrote:


Merged #7487<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_NREL_EnergyPlus_pull_7487&d=DwMCaQ&c=OCIEmEwdEq_aNlsP4fF3gFqSN-E3mlr2t9JcDdfOZag&r=rxT5qGz4Fgu2V4OupeO-GHp4q_5AzQNUBLnrS4uzhd8&m=-7Ku3JFXPvAfnaCQf43bjYTWlqT9OU1VU79KBo_vE2U&s=fIx1WCf14pbXldIHe0mqIKUtQy9JbCpQsU1b5c5Sp6c&e=> into develop.

—
You are receiving this because you were assigned.
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_NREL_EnergyPlus_pull_7487-3Femail-5Fsource-3Dnotifications-26email-5Ftoken-3DABMJHPLOFNQKPA2GUGSH4DTQKLLGLA5CNFSM4ITTI3E2YY3PNVWWK3TUL52HS4DFWZEXG43VMVCXMZLOORHG65DJMZUWGYLUNFXW5KTDN5WW2ZLOORPWSZGOTWWZDZI-23event-2D2645397989&d=DwMCaQ&c=OCIEmEwdEq_aNlsP4fF3gFqSN-E3mlr2t9JcDdfOZag&r=rxT5qGz4Fgu2V4OupeO-GHp4q_5AzQNUBLnrS4uzhd8&m=-7Ku3JFXPvAfnaCQf43bjYTWlqT9OU1VU79KBo_vE2U&s=cXQcdMYy-YuDNDkJRiGWLNfy2GaYmbKOaERDX5HxEAc&e=>, or mute the thread<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_ABMJHPNH2ZZNERZ2MQYSFQLQKLLGLANCNFSM4ITTI3EQ&d=DwMCaQ&c=OCIEmEwdEq_aNlsP4fF3gFqSN-E3mlr2t9JcDdfOZag&r=rxT5qGz4Fgu2V4OupeO-GHp4q_5AzQNUBLnrS4uzhd8&m=-7Ku3JFXPvAfnaCQf43bjYTWlqT9OU1VU79KBo_vE2U&s=2JLu3zQMqhnE9yfSLrvOhFv936NtFl2hbs0yuKD77Xc&e=>.
